### DISCUSSION TITLE
How to bind keys only for History Menu?

### DISCUSSION THREAD

**Discussion by JerryZhongJ** [marlonrichert/zsh-autocomplete#678] [Q&A] [answered]
I want to keep the default behaviors in zsh' history search, which is using `←` and `→` to quit selection and move forward/backward a char in command, and pressing 'Enter' to run the command in the menu immediately. For now, when I find out a command from history menu, I need a double `Enter` to run it, which is quite inconvenient. If I want to modify the command, I need to press `tab` or `Enter` to quit the menu first, which I always forget, and this brings me a mess.

I tried the solution in README.md, but this does not only modify the key binding in History Menu, but also in Completion Menu.
```
bindkey -M menuselect '\r' .accept-line
bindkey -M menuselect "$terminfo[kcub1]" .backward-char
bindkey -M menuselect "$terminfo[kcuf1]" .forward-char      
```
But this is not what I want. When I press `Enter` in Completion Menu, what I mean is to select the item.  And I can't move left and right among the items.

I know I can change the behavior of `↑` to the default of zsh, but I want to keep the History Menu because it is quite useful for me, 

I think the solution is to change key binding only for History Menu. Can it be done?

**Comment by JerryZhongJ** [comment, marked as answer]
I figure out a solution, that is adding a wrapper to the two widgets, history-search and complete-word. The wrappers call `bindkey` every time the widgets are called.

I add this to my `.zshrc`:
```
wrap(){
  functions[${1}:old]="$(
      unfunction ${1} 2> /dev/null
      builtin autoload +X -Uz ~autocomplete/Functions/Widgets/${1}
      print -r -- "$functions[${1}]"
  )"
  eval "${1}(){
    bindkey -M menuselect $terminfo[kcub1] ${2}
    bindkey -M menuselect $terminfo[kcuf1] ${3}
    bindkey -M menuselect '\r' ${4}
    ${1}:old \$@
  }"
}
wrap .autocomplete__history-search__completion-widget .backward-char .forward-char .accept-line
wrap .autocomplete__complete-word__completion-widget backward-char forward-char accept-line
unfunction wrap
```
However this solution is too ugly, and will break if the function names are changed. I'm looking forward to a better solution.

**Reply by marlonrichert** [reply]
I used to have code like this inside the history widgets, but the problem is that the bindings don't get reversed when, for example, you abort the menu with `^C`. Then you suddenly end up with these bindings in the completion menu, too, and that's not what you want.

**Reply by JerryZhongJ** [reply]
@marlonrichert This is not going to reverse the key binding when you quit the menus, but reset it every time you enter the menus.

**Reply by dingyaguang117** [reply]
Thank you very much! I think this should be the default action.

**Comment by theerror** [comment]
Yes, this is exactly what case me trouble as well. I've come to zsh from bash and I do like this history, BUT exactly as you told the needs of confirming command before I'm able to edit it somehow is really pain and switching next / previous one, when I hit the left/right and my intended command disappears, it drives me creasy too. This definitely makes sense to map like this by default (end history / search and start moving cursor in command prompt).

**Reply by DreadPirateFlint** [reply]
My people!  I figured I was the only person in the world so annoyed by such a very specific thing!  Tried to explain it to my wife and she just looked at me like 'why did I marry this person'.
